### PR TITLE
Stucture change for baseclass

### PR DESCRIPTION
**Pull request by jjmachan**
#### _Why is this change required? What problem does it solve? Describe your changes in detail:_
Proposes a small structure change in base class which will make it easier for adding datasets with different types of splits.
Also made a patch for transpose operation

## Types of changes
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Documentation update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: stockroom/external/importer/torchvision_importers.py [hunk 1]
```diff
@@ -41,26 +41,61 @@ def variability_status(self):
...
-        dataset = datasets.FashionMNIST(root=root, train=train, download=True)
-        super().__init__(dataset, train)
+    @classmethod
+    def gen_splits(cls, root):
+        dataset = datasets.FashionMNIST
+        dataset_splits = super().gen_splits(dataset, root)
+        return dataset_splits
+
+
+class ImageFolder(BaseImporter):
```
Comment (hhsecond) [inline stockroom/external/importer/torchvision_importers.py:91]: It might be better to submit as another PR when the feature is complete?

File: stockroom/external/importer/torchvision_importers.py [hunk 2]
```diff
@@ -26,7 +26,7 @@ def dtypes(self):
     @staticmethod
     def _process_data(img, lbl):
         # TODO: memory copy
-        img = np.ascontiguousarray(np.transpose(np.array(img), (2, 0, 1)))
+        img = np.ascontiguousarray(np.array(img))
```
Comment (hhsecond) [inline stockroom/external/importer/torchvision_importers.py:29]: you don't need `np.ascontiguousarray` here. The array is already contiguous

File: stockroom/external/importer/torchvision_importers.py [hunk 3]
```diff
@@ -41,26 +41,61 @@ def variability_status(self):
...
+    @classmethod
+    def gen_splits(cls, root):
+        dataset = datasets.CIFAR10
+        dataset_splits = super().gen_splits(dataset, root)
+        return dataset_splits
+
+    @staticmethod
+    def _process_data(img, lbl):
+        img = np.transpose(np.array(img), (2, 0, 1))
+        img = np.ascontiguousarray(np.array(img))
```
Comment (hhsecond) [inline stockroom/external/importer/torchvision_importers.py:65]: You don't need the inner `np.array` in this line. It's already an `np.ndarray` object

File: stockroom/external/importer/torchvision_importers.py [hunk 4]
```diff
@@ -41,26 +41,61 @@ def variability_status(self):
...
 
-    def __init__(self, root, train=True):
-        dataset = datasets.FashionMNIST(root=root, train=train, download=True)
-        super().__init__(dataset, train)
+    @classmethod
+    def gen_splits(cls, root):
+        dataset = datasets.FashionMNIST
+        dataset_splits = super().gen_splits(dataset, root)
+        return dataset_splits
+
```
Comment (hhsecond) [inline stockroom/external/importer/torchvision_importers.py:89]: `FashionMnist` also need to override the `_process_data` from the base class, right?

File: stockroom/external/importer/torchvision_importers.py [hunk 5]
```diff
@@ -41,26 +41,61 @@ def variability_status(self):
...
 class Cifar10(TorchvisionCommon):
     name = 'cifar10'
 
-    def __init__(self, root, train=True):
-        dataset = datasets.CIFAR10(root=root, train=train, download=True)
-        super().__init__(dataset, train)
+    @classmethod
+    def gen_splits(cls, root):
+        dataset = datasets.CIFAR10
+        dataset_splits = super().gen_splits(dataset, root)
```
Comment (hhsecond) [inline stockroom/external/importer/torchvision_importers.py:59]: Perhaps, make it a one-liner? Also call `cls.get_splits` instead of `super()`?
```python
return cls.gen_splits(datasets.CIFAR10, root)
```

Both are not deal-breaking problems. I think it's more readable

File: stockroom/external/importer/torchvision_importers.py (hunk 1 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:91]: sure

File: stockroom/external/importer/torchvision_importers.py (hunk 4 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:89]: is it required? Since process is a static method its already in the namespace when we inherit rt?
I'm not sure since I not clear about how inheritance of static method works but the code is calling the `_process_data`  of the parent and I wrote the parent `_process_data` to handle mnist and fashionMnist

File: stockroom/external/importer/torchvision_importers.py (hunk 5 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:59]: Yes I agree, this makes it a lot more readable

Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:59]: this should be 
``` python
  return super().gen_splits(datasets.CIFAR10, root)
```
rt?
cls will refer to the current class but we want to call the method from the base class rt?
is this how it works?

File: stockroom/external/importer/torchvision_importers.py (hunk 2 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:29]: yep, changed

File: stockroom/external/importer/torchvision_importers.py (hunk 4 above)
Comment (hhsecond) [reply stockroom/external/importer/torchvision_importers.py:89]: I am sorry. You are right. I was in an assumption that `FashionMnist` has color channels.

File: stockroom/external/importer/torchvision_importers.py (hunk 5 above)
Comment (hhsecond) [reply stockroom/external/importer/torchvision_importers.py:59]: make sense

File: stockroom/external/importer/torchvision_importers.py (hunk 3 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:65]: when I checked it `img` is a `pill.Image` object. Since we are overwriting the base method we have to convert it rt?
I tried to call the method in the base class but that seems not very neat since its a static method hence I thought of implementing it again.

Comment (hhsecond) [reply stockroom/external/importer/torchvision_importers.py:65]: `img` is a `pil.Image` but that is converted to an np array on line 64. On line 65, you are calling `np.array` again

File: stockroom/external/importer/torchvision_importers.py (hunk 4 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:89]: Oh yeah got it. My bad :sweat_smile: 
fixed it!

File: stockroom/external/importer/torchvision_importers.py (hunk 3 above)
Comment (jjmachan) [reply stockroom/external/importer/torchvision_importers.py:65]: Oh I didn't see that. My bad :sweat_smile: 
fixed it!
